### PR TITLE
Feat: Add support for exporting and querying flowNodes in RDBMS #23268

### LINKED ISSUE #23268: Implement Exporter and SearchClient for FlowNodeInstance
The RDBMS Camunda 8 REST API V2 should support querying for FlowNodeInstance.

**AcceptanceCriteria:**
- the RDBMS module fully implements the FlowNodeInstanceSearchClient with all filters and fields
- the RDBMS module can process all relevant Zeebe records for the FlowNodeInstance

### PR DESCRIPTION
**Pull request by stoerti**
## Description

Adds support to search for flowNodes via SearchAPI.

## Related issues
- TreePath and incident fields will be covered in [#23270](https://github.com/camunda/camunda/issues/23270)
- ProcessInstance and FlowNode migration will be covered in [#24057](https://github.com/camunda/camunda/issues/24057)

closes #23268

### CODE REVIEW

File: db/rdbms/src/main/java/io/camunda/db/rdbms/write/queue/ExecutionQueue.java [hunk 1]
```diff
@@ -25,7 +26,8 @@ public class ExecutionQueue {
   private final List<PreFlushListener> preFlushListeners = new ArrayList<>();
   private final List<PostFlushListener> postFlushListeners = new ArrayList<>();
 
-  private final Queue<QueueItem> queue = new ConcurrentLinkedQueue<>();
+  // TODO solve concurrency
```
[the remark this answers is not in the document]
Comment (stoerti) [reply db/rdbms/src/main/java/io/camunda/db/rdbms/write/queue/ExecutionQueue.java]: I extracted all this to a new issue/branch [#24051](https://github.com/camunda/camunda/issues/24051)

File: db/rdbms/src/main/java/io/camunda/db/rdbms/write/service/FlowNodeInstanceWriter.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    executionQueue.executeInQueue(
+        new QueueItem(
+            ContextType.FLOW_NODE,
+            flowNode.flowNodeInstanceKey(),
+            "io.camunda.db.rdbms.sql.FlowNodeInstanceMapper.insert",
+            flowNode));
+  }
+
+  public void update(final FlowNodeInstanceDbModel flowNode) {
+    executionQueue.executeInQueue(
```
[the remark this answers is not in the document]
Comment (stoerti) [reply db/rdbms/src/main/java/io/camunda/db/rdbms/write/service/FlowNodeInstanceWriter.java:41]: Extracted to [#24051](https://github.com/camunda/camunda/issues/24051)

File: db/rdbms/src/main/java/io/camunda/db/rdbms/write/service/FlowNodeInstanceWriter.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    @Override
+    public boolean canBeMerged(final QueueItem queueItem) {
+      return queueItem.contextType() == ContextType.FLOW_NODE
+          && queueItem.id().equals(dto.flowNodeKey())
+          && queueItem.parameter() instanceof FlowNodeInstanceDbModel;
+    }
+
+    @Override
+    public QueueItem merge(final QueueItem originalItem) {
+      final var newParameter =
```
[the remark this answers is not in the document]
Comment (stoerti) [reply db/rdbms/src/main/java/io/camunda/db/rdbms/write/service/FlowNodeInstanceWriter.java]: Extracted to [#24051](https://github.com/camunda/camunda/issues/24051)

File: zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/ProcessInstanceExportHandler.java [hunk 4]
```diff
@@ -35,13 +51,17 @@ public boolean canExport(final Record<ProcessInstanceRecordValue> record) {
   public void export(final Record<ProcessInstanceRecordValue> record) {
     if (record.getValue().getBpmnElementType() == BpmnElementType.PROCESS) {
       exportProcessInstance(record);
+    } else {
+      if (FLOW_NODE_INTENT.contains(record.getIntent())) {
```
[the remark this answers is not in the document]
Comment (stoerti) [reply zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/ProcessInstanceExportHandler.java]: Did that. RdbmsExport now defines a List for every type.

File: zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/ProcessInstanceExportHandler.java [hunk 5]
```diff
@@ -73,7 +93,38 @@ private void exportProcessInstance(final Record<ProcessInstanceRecordValue> reco
...
+          record.getKey(),
+          FlowNodeState.COMPLETED,
+          DateUtil.toOffsetDateTime(Instant.ofEpochMilli(record.getTimestamp())));
+    } else if (record.getIntent() == ProcessInstanceIntent.ELEMENT_TERMINATED) {
+      flowNodeInstanceWriter.update(
+          record.getKey(),
+          FlowNodeState.TERMINATED,
+          DateUtil.toOffsetDateTime(Instant.ofEpochMilli(record.getTimestamp())));
+    } else if (record.getIntent() == ProcessInstanceIntent.ELEMENT_MIGRATED) {
+
```
[the remark this answers is not in the document]
Comment (stoerti) [reply zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/ProcessInstanceExportHandler.java]: Extracted the ProcessInstance migration to a new issue [#24057](https://github.com/camunda/camunda/issues/24057)

File: zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/ProcessInstanceExportHandler.java [hunk 6]
```diff
@@ -73,7 +93,38 @@ private void exportProcessInstance(final Record<ProcessInstanceRecordValue> reco
     }
   }
 
-  private ProcessInstanceDbModel map(final Record<ProcessInstanceRecordValue> record) {
+  private void exportFlowNode(final Record<ProcessInstanceRecordValue> record) {
+    final var value = record.getValue();
+    if (record.getIntent() == ProcessInstanceIntent.ELEMENT_ACTIVATING) {
```
[the remark this answers is not in the document]
Comment (stoerti) [reply zeebe/exporters/rdbms-exporter/src/main/java/io/camunda/exporter/rdbms/ProcessInstanceExportHandler.java]: The same Intent is used in CamundaExporter as well. So I would keep it aligned to the other exporters

**Review by danielkelemen — APPROVED** [review]
Looks good, added a question about column sizing.

File: db/rdbms/src/main/resources/db/changelog/rdbms-support/changesets/create_engine.xml [hunk 7]
```diff
@@ -76,4 +76,31 @@
     </createIndex>
   </changeSet>
 
+
+  <changeSet id="create_flow_node_table" author="cthiel">
+    <createTable tableName="FLOW_NODE_INSTANCE">
+      <column name="FLOW_NODE_INSTANCE_KEY" type="BIGINT">
+        <constraints primaryKey="true"/>
+      </column>
+      <column name="FLOW_NODE_ID" type="VARCHAR(400)" />
```
[the remark this answers is not in the document]
Comment (danielkelemen) [reply db/rdbms/src/main/resources/db/changelog/rdbms-support/changesets/create_engine.xml:86]: Good question, I'm also not sure what valid size in C8 could be. I guess they are generated somewhere in zeebe with an algorithm, that could have a max size.
C7 is either using 255 or 64, but mostly 64, that's already quite long.

**Pull request merged**